Expose server config, session queue and node as MBeans for JMX monitoring.

pujagani opened the pull request:
### Description
<!--- Describe your changes in detail -->
Changes are related to #7761.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Old Grid exposed MBeans for JMX monitoring. The changes include exposing MBeans and attributes for monitoring the new Grid. This includes:
1. Exposing BaseServerConfig and related attributes.
2. Exposing Node and related attributes.
3. NewSessionQueue and related attributes.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

AutomatedTester commented on java/server/src/org/openqa/selenium/grid/server/BaseServerOptions.java:
```diff
@@ -94,6 +104,15 @@ public URI getExternalUri() {
+    try {
+      return getExternalUri().toURL();
+    } catch (MalformedURLException e) {
+      throw new ConfigException("Cannot convert URI to URL" + e.getMessage());
```
nit: We probably want a space after `URL`

diemol commented on java/server/src/org/openqa/selenium/grid/sessionqueue/local/LocalNewSessionQueue.java:
```diff
@@ -80,6 +88,17 @@ public boolean isReady() {
+  @ManagedAttribute(name = "NewSessionQueueSize")
+  public int getQueueSize() {
+    Lock writeLock = lock.writeLock();
+    writeLock.lock();
```
Why do we need to lock to read the queue size?

pujagani replied:
Since it uses ConcurrentLinkedDeque, the size() requires a traversal and value may not be accurate if it is being modified when size is calculated. Hence the lock. But for exposing values if we don't need that kind of accuracy, then we can get rid of the lock. @diemol What would you suggest?

diemol reviewed (changes requested):
👋 @pujagani, I went through it and I added a comment besides the one already exists.

Also, it seems tests for this PR are not passing, and a couple are still to be written. Could you please have a look? I'll happily review right after.

pujagani commented:
@diemol I updated and added the tests and they all pass now. The test was picking up previously registered MBeans and hence failing for different cases. Please help review. Thank you!

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/local/LocalNewSessionQueue.java:
```diff
@@ -80,6 +88,17 @@ public boolean isReady() {
 
+  @ManagedAttribute(name = "NewSessionQueueSize")
+  public int getQueueSize() {
+    Lock writeLock = lock.writeLock();
```
This need only be a read lock

shs96c commented on java/server/src/org/openqa/selenium/grid/node/local/LocalNode.java:
```diff
@@ -181,11 +189,53 @@ public boolean isReady() {
+  }
+
+  @ManagedAttribute(name = "RemoteNodeUri")
+  public URI getExternalURI() {
```
Nit: `getExternalUri` We try and AVOID SHOUTING in Selenium (which is why we have `HttpMethod`, and so on)

shs96c commented on java/server/src/org/openqa/selenium/grid/server/BaseServerOptions.java:
```diff
@@ -101,6 +111,15 @@ public URI getExternalUri() {
   }
 
+  @ManagedAttribute(name = "URL")
+  public URL getExternalUrl() {
```
Daft question, but why don't we just return the URI?

shs96c commented on java/server/test/org/openqa/selenium/grid/router/JmxTest.java:
```diff
@@ -0,0 +1,222 @@
+
+      String retryInterval = (String) beanServer.getAttribute(name, "RetryIntervalSeconds");
+      assertThat(Long.parseLong(retryInterval)).isEqualTo(queueOptions.getRetryIntervalSeconds());
+    } catch (InstanceNotFoundException | IntrospectionException | ReflectionException
```
It's fine to just let the test fail because the exception was thrown. That'll give you more information that just calling `fail` in these `catch` blocks.

diemol reviewed (approved):
Thank you, @pujagani!
